### DISCUSSION TITLE
WiFi-RTU_Bridge example help

### DISCUSSION THREAD

**Discussion by KeystoneJack** [eModbus/eModbus#332] [Q&A] [answered]
Hi!

I'm trying to get the WiFi-RTU_Bridge example to work on my Lolin D32. After I've uploaded the sketch and check the Serial Output all I get is 
`__ OK __
.IP addr�` 
and nothing more...  If I uncomment this row 
` Serial2.begin(9600, SERIAL_8N1, GPIO_NUM_3, GPIO_NUM_1);` 
I get more output. If I don't use the `Serial2` at all I get the correct output. My Arduino and cpp skills are at beginner level so I'm probably doing something wrong.

Basically what I want to do is to relay all TCP traffic to RTU. 

Here is my setup:
Hardware:
Lolin D32 
DSD-tech SH-U12 

Wiring:
LOLIN TX0 <---> RXD SH-U12
LOLIN RX0 <---> TXD SH-U12
LOLIN GND <---> GND SH-U12
LOLIN 3V <---> 5VO SH-U12 (Works with 3.3V as well)

Any help is appreciated!

**Comment by Miq1** [comment]
The first log line is not showing any IP address? Then something with your WiFi connection may be failing. You should first try to get the ESP to successfully connect to your network before fiddling with something else. 

The example sketch will output nothing else but the IP and the servers attached if not requested via TCP (like is shown in the comments at the end of the example sketch).

There may be terminating/biasing issues with the RTU adapter as well, but we will see to that in a second step,

**Reply by KeystoneJack** [reply]
It connects to my wifi! Only it doesn't print the entire ip. Which is strange!

**Reply by KeystoneJack** [reply]
This also occurs when I unplug SH-U12.
I will test another Esp32 tomorrow and report back!

**Reply by Miq1** [reply]
Strange indeed. The output `` .IP addr�`` tells us the state was ``WL_CONNECTED`` after the first attempt, but the first byte of the IP is zero?
I am not convinced another ESP will behave differently, but we will see.

**Reply by KeystoneJack** [reply]
Oddly enough it actually work with another ESP... I haven't hooked it up to my SH-U12 yet. Will do that later but at least it prints:
`__ OK __
...IP address: 10.1.1.12
[N] 3877| ModbusServer.cpp     [ 163] listServer: Server   4:  00
Use the shown IP and port 502 to send requests!
`
Which is good!

**Reply by Miq1** [reply]
Good news. I can not imagine a way to prevent some ESP to connect and *not* get an IP address, though.

**Reply by KeystoneJack** [reply]
Me neither! 

Anyway, the new ESP is now hooked up to my SH-U12 and my SH-U12 to my heating systems central unit, which has a modbus port. 
I can connect to the ESP over TCP perfectly but I can't read anything from it. All the responses are empty. Which probably suggest I have hooked the SH-U12 wrong... 

The new ESP is a Lolin C3 mini and I'm using the RX TX pins. Do you know if it's correct? 
I'm also using `Serial1`instead of `Serial2` because `Serial2` didn't work.

Thanks for your help!

**Reply by Miq1** [reply]
I did not have a sample yet of that ESP variety, but RX/TX are reasonable, as is Serial1.

"Empty" responses strictly speaking are not possible with eModbus, you at least should get some error response like TIMEOUT. What does the bridge's log show?

- You will of course have crossed RX and TX between the ESP and the adapter?
- What is the length of the RS485 wires? If longer than a few cms you will have to bias the bus: put a 4k7 resistor each between A+ of the bus to Vcc and between B- and GND.

**Reply by KeystoneJack** [reply]
I have crossed the RX TX so RX goes to TX and vice versa.
The length of the wires are 10cm.

How do I enable logging?
I've tried 
`#define LOG_LEVEL LOG_LEVEL_VERBOSE
#include <Logging.h>`
but that did nothing...

**Reply by Miq1** [reply]
Are you using the Arduino IDE? If so, you will have to change the default ``LOG_LEVEL`` by editing ``Logging.h``.
I meant the RS485 wires, by the way. The UART wires are regardless.

**Reply by KeystoneJack** [reply]
I am.

Yes, the RS485 wires are 10 cm.

**Reply by Miq1** [reply]
And...? More log lines now?

**Reply by KeystoneJack** [reply]
Haven’t tries yet. Ran out of time. Will try later.

I have a RS485 usb-dongle so I can troubleshoot better. Any tips on a RTU-simulator would be appreciated 😊

**Reply by Miq1** [reply]
I have none... Most of the time I will use my real-world bus in the basement 🤣 

Looking forward to see the log.

**Reply by KeystoneJack** [reply]
Before I test and post more logs. Is it possible that my SH-U12 with MAX13487 chip is not supported?

**Reply by Miq1** [reply]
I would not think so. It has all that is required, and, as a bonus, is auto-half duplex.

It is specified for 5V, so you may have to use a level shifter.

**Reply by KeystoneJack** [reply]
Alright! I have some logs. 
Here I'm trying to read a coil on address 1000

```
[D] 29638| ModbusBridgeTemp.h   [ 174] bridgeWorker: Request (01/01) sent 
[D] 29638| ModbusClientRTU.cpp  [ 230] addToQueue: RC=01. 
[D] 29638| ModbusClientRTU.cpp  [ 248] handleConnection: Pulled request from queue
[D] Sent packet: @3FCAB7B4/6:
  | 0000: 01 01 03 E8 00 01                                 |......          |
[D] 29647| ModbusClientRTU.cpp  [ 253] handleConnection: Request sent.
[D] 31648| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FCAB9C4/1:
  | 0000: E0                                                |.               |
[D] 31649| ModbusClientRTU.cpp  [ 268] handleConnection: Error response (1 bytes) received.
[D] 31649| ModbusClientRTU.cpp  [ 289] handleConnection: Response generated.
[D] 31658| ModbusServerTCPtemp.h [ 314] worker: Data response
[D] 32260| ModbusServerTCPtemp.h [ 355] worker: Worker stopping due to timeout.

```
Which seems correct right?

**Reply by KeystoneJack** [reply]
To add. This is when my SH-U12 isn't hooked up to my RTU

**Reply by KeystoneJack** [reply]
And here is the log when it's hooked up:
```
[D] 63343| ModbusServerTCPtemp.h [ 210] accept: Started client 0 task 1070249504
[D] 63344| ModbusServerTCPtemp.h [ 239] serve: Accepted connection - 1 clients running
[D] 63344| ModbusServerTCPtemp.h [ 261] worker: Worker started, timeout=600
[D] 63345| ModbusServer.cpp     [  38] getWorker: Worker found for 01/ANY
[D] 63345| ModbusBridgeTemp.h   [ 174] bridgeWorker: Request (01/01) sent
[D] 63345| ModbusClientRTU.cpp  [ 230] addToQueue: RC=01
[D] 63345| ModbusClientRTU.cpp  [ 248] handleConnection: Pulled request from queue
[D] Sent packet: @3FCAB8B8/6:
  | 0000: 01 01 03 E8 00 01                                 |......          |
[D] 63354| ModbusClientRTU.cpp  [ 253] handleConnection: Request sent.
[D] 65355| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FCABAC8/1:
  | 0000: E0                                                |.               |
[D] 65356| ModbusClientRTU.cpp  [ 268] handleConnection: Error response (1 bytes) received.
[D] 65356| ModbusClientRTU.cpp  [ 289] handleConnection: Response generated.
[D] 65364| ModbusServerTCPtemp.h [ 314] worker: Data response
[D] 65967| ModbusServerTCPtemp.h [ 355] worker: Worker stopping due to timeout.
```

**Comment by Miq1** [comment]
Great. You are getting E0 TIMEOUT error responses from the lib, nothing from the server at all. Some common reasons are:

- your wiring has a problem - try switching RX and TX. A+/B- connected on both sides of the bus to the respective terminals?
- the adapter has a power issue
- your RS485 is lacking proper biasing
- the server has a different server ID than you thought it has
- the server does not honor the 01 function code, but is set to not send any error response
- the server takes more than 2s to respond (unlikely)

**Reply by KeystoneJack** [reply]
Good to hear! 
Then I guess this is down to, in network terms, a layer 8 problem 😂.
This is now out of scope for this eModbus-project and I really appreciate your help. Thank you!

Just for info, what hardware are you using? What esp and what adapter. Would be fun to test with a set of hardware that we know work.

**Reply by Miq1** [reply]
I honestly would doubt the ESP being the root cause. We had numerous different varieties known to be working, but issues with RTU are legion instead. Do you have a scope available you could hook up to the A+ and B- bus wires?

BTW, my personal ESP most used is the Devkit C V4 type, as my projects mostly are needing a substantial number of GPIOs.

**Reply by KeystoneJack** [reply]
I agree!
I’m currently using a raspberry pi , running a python script,with a usb dongle,  which works great.

I don’t have an oscilloscope unfortunately…

**Reply by Miq1** [reply]
The log tells us there is no data on the UART side of the adapter, so the problem must be located in or beyond the adapter. With a multimeter you could measure the voltage between A and B, with a real fast sampling one you should even be able to see the levels.

**Reply by KeystoneJack** [reply]
More troubleshooting!

- Tried multimeter between A and B on the SH-U12. Got faint reading, which I guess is correct.
- Changed cables, didn't work. Still get error.

With my USB adapter hooked up to my raspberry pi, using the EXACT same cable as with my Arduino it works. 
The way I test this is with pymodbus REPL. 
 My modbus client has these settings:
- Modbus address is adjustable between 1 - 247. Default: 1  
- The baud rate is adjustable to 9600, 19200 or 38400 baud. Default: 38400  
- Parity is adjustable to: None, Odd, Even. Default: No 
- Stop Bits is adjustable to 1 or 2. Default: 1 
- Data Bits: 8

and here is my Arduino script:
```
#include <Arduino.h>
#include "HardwareSerial.h"
#include <WiFi.h>
#define LOG_LEVEL LOG_LEVEL_VERBOSE
#include <Logging.h>

#include "ModbusBridgeWiFi.h"
#include "ModbusClientRTU.h"

#ifndef MY_SSID
#define MY_SSID "SomeWifi"
#endif
#ifndef MY_PASS
#define MY_PASS "VeryS3cret"
#endif

char ssid[] = MY_SSID;
char pass[] = MY_PASS;
uint16_t port = 502;

ModbusClientRTU MB;

ModbusBridgeWiFi MBbridge;

void setup() {
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

  RTUutils::prepareHardwareSerial(Serial1);

  WiFi.begin(ssid, pass);
  delay(200);
  while (WiFi.status() != WL_CONNECTED) {
    Serial.print('.');
    delay(1000);
  }
  IPAddress wIP = WiFi.localIP();
  Serial.println(WiFi.status());
  Serial.printf("IP address: %u.%u.%u.%u\n", wIP[0], wIP[1], wIP[2], wIP[3]);
  Serial1.begin(9600, SERIAL_8N1, GPIO_NUM_3, GPIO_NUM_1);

  MB.setTimeout(5000);
  MB.useModbusRTU();
  MB.begin(Serial1);

  MBbridge.attachServer(1, 1, ANY_FUNCTION_CODE, &MB);

  MBbridge.listServer();


  MBbridge.start(port, 4, 5000);

  Serial.printf("Use the shown IP and port %d to send requests!\n", port);
}

void loop() {
  delay(1000);
}
```

**Reply by Miq1** [reply]
The sketch looks okay; I suppose your server's serial is set to 9600 8N1 as well? 

So it is pointing again to the adapter and wires. 

Are you powering the adapter with 5V or 3V3? It may be the adapter will not work correctly with lower voltage than it is specified. But you will have to put voltage dividers or a proper level shifter to RX and TX to not fry the ESP's GPIOs with 5V signals!

Second suggestion would be to add the mentioned biasing resistors to the RS485. May be the adapter will need it although the USB dongle does work without.

**Reply by KeystoneJack** [reply]
> I suppose your server's serial is set to 9600 8N1 as well?

The server in this case is it my ESP?

**Reply by Miq1** [reply]
No, I meant the Modbus server you are going to request on the RTU bus. On the RTU bus the ESP as a bridge is the Modbus client. Its TCP side again is a Modbus server, but that is irrelevant here. We are looking at the RTU bus only.

**Reply by KeystoneJack** [reply]
I don't know...
Here is my python script which runs on my pi and is working:
```
import asyncio
import logging

from pymodbus.client import ModbusSerialClient
from pymodbus.datastore import ModbusServerContext
from pymodbus.datastore.remote import RemoteSlaveContext
from pymodbus.server import StartAsyncTcpServer

_logger = logging.getLogger(__file__)


async def run_forwarder(port, slaves=None):
    """Run forwarder setup."""
    txt = f"### Start forwarder, listen {port}, connect to {port}"
    _logger.info(txt)

    client = ModbusSerialClient(port="/dev/ttyACM0", baudrate=38400)
    client.connect()
    assert client.connected

    if slaves:
        store = {i.to_bytes(1, "big"): RemoteSlaveContext(client, slave=i) for i in slaves}
    else:
        store = RemoteSlaveContext(client, slave=1)

    context = ModbusServerContext(slaves=store, single=True)
    await StartAsyncTcpServer(context=context, address=("", port))


if __name__ == "__main__":
    # Example usage
    server_port = 5020

    asyncio.run(run_forwarder(port=server_port))

```

The baud rate is set to 38400, so I can try that with my esp as well. If that don't work I'll think I'll stick to my raspberry pi unfortunately... I'm way out of my modbus comfort zone 😂

**Reply by Miq1** [reply]
OMG. The baud rate has to be that of the server and if your Raspi does work ok with 38400 it is unlikely the ESP will work at 9600...

**Reply by KeystoneJack** [reply]
How can I be so stupid… 🤦‍♂️
The docs clearly states that the default is 38400… Sigh… Sorry @Miq1

**Reply by KeystoneJack** [reply]
So... I have changed the baud rate to 38400 and guess what? I didn't work... Which means I'll stick to my pi!

**Reply by Miq1** [reply]
Nah! No way! Do not give up so easily! 😉 

Can you take a photo of your ESP and RS485 Adapter and the wiring? It may give another aspect on what is going on.

**Reply by KeystoneJack** [reply]
I use VBUS in order to supply 5V

**Reply by Miq1** [reply]
Yeah, and that brings the RX and TX signals to 5V as well... You may have fried the UART with that, the ESP's GPIOs are 3V3-tolerant only. As I wrote: get yourself a level shifter and plug it in between the RX and TX lines.

**Reply by KeystoneJack** [reply]
I see... I'll try that! Will get back to you in a couple of days. Thanks so far!

**Reply by KeystoneJack** [reply]
I got a hold of an esp8266 that has 5v output. Is there any trick to get it running on an esp8266?

**Reply by Miq1** [reply]
No, unfortunately the bridge code needs an ESP32, due to the multitasking design. The ESP8266 has a single user task only.

And as far as I remember the ESP8266 is 3V3-tolerant only as well. A level shifter or 4 resistors for two voltage dividers at the RX and TX lines will be the better solution.

**Reply by KeystoneJack** [reply]
I hear you! Purchased a level shifter, just waiting for it to arrive

**Reply by KeystoneJack** [reply, marked as answer]
Many maaaaany days later and it's now working.
Huge thanks to @Miq1 for the help!

The solution was to buy another esp32 and a new RS485 module.

**Reply by Miq1** [reply]
👍🏻

**Reply by KeystoneJack** [reply]
A more detailed response. 
It might of worked with the 5V level shifter but I'm too stupid to get it to work. Buying a simpler RS485 unit that directly connects to an ESP32 via I2C is much easier on my knowledge.
